### ISSUE TITLE
1841 Implementation Checklist

### ISSUE THREAD

**Issue by roseundy** [tobymao/18xx#7623]
### BHAGs
**Passes**
* A new revenue center type, most likely a new class derived from Engine::Part::City
* They would require a cost parameter, possibly also a color
* They would render as a standard city (including slots) over a filled triangle
* Don't count towards the requirement of having two revenue centers in a route (neither do ports)
* A token in a pass doesn't count towards the requirement of having a tokened city in a route
* Passes count towards the train limit - except for 8 trains

**Regions/Borders**
* Changes depending on phase
* Cannot lay across a border in phase 2
* When determining legal token locations, route from existing token can never cross a border. Change to Graph? How?
* Merging corporations must have a route between them (connecting tokens) that doesn't cross a border
* Replacing another company's token requires a route to an existing token that doesn't cross a border.
* Train routes may cross borders
* New 18xx.games border type?

**Graph from each token**
* Some phases allow multiple tiles to be laid - but each must trace from different existing tokens
* Modify Graph? Separate graphs?

**Token Costs**
* Tokens are purchased when a corporation starts
* After selecting the one or two bases for initial token placement, the cost is determined
* Per token cost is fixed for historical corporations
* For non-historical corporations, the cost is a function of the proximity of connected track to the token
* For the above, track across from an active border and non-connected pre-printed track are ignored
* If the cost cannot be paid, the corporation undergoes EMR
* The rules seem to imply home tokens are placed before purchase, since, at least for non-historical corporations, the token cost is determined by those placements.
* To implement, mandatory number of tokens (2 for major, 1 for minor) will pre-exist in corporations and tokens beyond those one or two will be added by the player.

**Cross-company token purchase**
* New step to replace a token from a willing company
* May be able to use Harzbahn code

**Chain of Control**
* concept: Controller. Always a player. For most games, always the President, but not in 1841
* controlled_corporations(player) - returns an unordered list of corporations that the player ultimately controls
* chain_of_control(corp) - returns an ordered list of corporation between corp and the controller plus the controller
* controller(corp) - returns Player that controls this corp, possibly via a chain
* How to detect/deal with circular chains

**Historical Corporations**
* President != biggest shareholder (until phase 4)
* Requires concession to form
* Concession+Presidency can be sold to another player. Cannot sell one w/o the other for an active corp
* Can sell inactive concession to another player by itself

**Frozen Corporations**
* Includes corps in a circular chain as well as corps in Receivership
* Rare, but possible
* Need examples

**Receivership**
* When does this happen exactly?

**Player to Player Stock Purchases**
* Cannot use usual "did you check with the other player" warning message
* Needs to use an offer/response mechanism similar to Rolling Stock. May have to modify UI to accomodate both

**Mergers**
* Similar to 1862 mergers, but less well defined

**Ferdinandea Secession**
* At start of phase 4 if there is no route between Milano and Venezia.
* Only way the SFL and SB (SFV for V1) Historical corporations enter the game
* Many rules

**Tuscan Merge**
* At start of phase 4 if at least two of SFLP, SFMA and SSFL are active.
* Only way the SFLi Historical corporation enters the game
* Many rules

**EMR**
* If corporation must purchase a train but cannot afford one, OR if corporation must puchase a token, but cannot afford one
* Starting with currently operating corporation, and the working up through controlling corporations until possibly reaching a player
* Must sell everything in treasury, then remaining IPO shares - all at half cost. Cannot sell a share if it causes a shift of presidency of operating corp or any controlling corp.  Corporations not in the chain of control may enter receivership. The bank pool is allowed to hold > 50% of a corporations shares and president shares during EMR. Also, shares for corps that haven't operated can be sold.
* If enough money is left over, additional trains may be bought. No additional tokens may be bought.
* If not enough money can be raised, all raised funds are given to the operating corporation and it doesn't have to buy a train this OR, and the remainder of its turn this OR is skipped. The player must sell all remaining shares for half price and declare bankruptcy. The operating company will either become frozen, or another player will become president and be on the hook next OR.

References:
* [Clarification on mergers](https://groups.io/g/18xx/topic/40525891#29878)

### Checklist
**Map**
- [x] Create new *Pass* part, inherited from city. Renders with triangle, has upgrade cost.
- [x] Tile list
- [x] Hexes
- [x] Is there a need for a *Port" part, since they are treated differently from towns?

**Entities**
- [x] Corporations (both major and minor)
- [x] Concessions (uses Companies)
- [x] Companies (Bayard) pay at end of each SR, not at the beginning of each OR

**Stock Market**
- [x] Mark area for major corporations only
- [x] Mark area for phase 8 only
- [x] Implement movement rules for above
- [ ] Multiple sales during turn affect stock prices in corporate turn order (not as they occur). I may blow this off unless someone complains.

**Track Laying**
- [x] Phase2: allow number of lays up to number of railheads (non-pass laid tokens)
- [x] Phase3,4: as above but max out at two
- [x] When laying multiple tiles per turn, each needs to trace to unique railhead
- [x] Phase2: cannot lay across a regional border
- [x] New track must connect to token in a city. Token in a pass doesn't count.

**Token Laying**
- [x] Can only lay tokens in cities that don't trace back to existing tokens across regional borders
- [x] Prevent any tokens in Firenze until SFMA starts or Phase 4

**Parring Corporations**
- [x] Phase 2: only historical corps
- [x] Phase 3+: can start non-historical corps
- [x] Allow purchase up to 40% of Corp when parring (only needs work for majors)
- [x] Auto token lays for historical corps
- [x] President selected token lays for SFMA and for non-historical corps
- [x] Constraints on location of non-historical inital token lays
- [x] Up to two token lays for majors
- [x] President selects # of tokens to buy (2-5 for majors, 1-2 for minors)
- [x] Tokens cost $50 for historical corps
- [x] For non-historical corps, token cost determined by proximity of chosen token locations to connected track

**Routing/Revenue Rules**
- [x] Route must contain tokened city and one other city, town, or off-board (passes and ports don't count)
- [x] Passes use up train "number" - except for 8 trains
- [x] Offboard values change on phase 4 (gray) and phase 6 (black)
- [x] Stock value only moves right if total dividend is greater than stock price

**Cross-company Token Purchase**
- [x] Need to add UI to allow setting of purchase price along with map and token selection
- [x] Add explicit confirmation step when across players?
- [x] Cannot purchase last city token from a corporation (i.e. only leave tokens in passes)

**Corporate Financial Operations**
- [x] Sell/Issue stock
- [x] Buy stock
- [x] Par corporations
- [x] Corps can only hold 5 non-IPO certs
- [x] Prevent circular ownership
- [x] Prevent human from controlling more than 60% of stock
- [x] Prevent purchase of controlling corp stock
- [x] Above rules also apply during stock rounds
- [x] Detect frozen corps on stock sales
- [x] Detect unfrozen corps on stock purchases
- [x] Show all controlled corps under proper player on Entities tab
- [x] Forced sales when over cert limit or when player controls more than 60% of stock
- [x] Line of presidential succession - when tied - goes to player in player order from the old president and then to corporations is operating order.
- [x] Deal with frozen historical corps prior to phase 4 by always ensuring the purchase of corresponding concession from the bank comes with the president cert (at an appropriate cost) 

**Concessions**
- [x] Initial Auction of concessions
- [x] Reorder before SR1
- [x] Lose un-exercised concessions at end of stock round (except for Bayard)
- [x] Can buy unused concessions from market, starting with SR2
- [x] Concessions required to par historical corps
- [x] Concessions prevent president changes
- [x] Concessions removed on Phase 4
- [x] Remove unstarted historical corporations on Phase 4

**Trains**
- [x] The seven 8T trains is a hard limit. Once those are gone, corporations are relieved the the requirement to own a train.
- [x] must_buy needs to take passes and ports into account

**Mergers**
- [x] Only if active corp has completed one OR.
- [x] Active president picks corp to merge with. Two corps must both be minors or both be majors. Two corps must have city tokens connected w/o passing through an active border. If historical corp is merging, must be phase 4 or later. Corp must have completed at least one OR.
- [x] Active president picks new corp to form (from unstarted non-historical major corps)
- [x] All assets from old corps are transferred to new corp
- [x] Any shares of old corps in new corp treasury (**not** IPO) are eliminated
- [x] Determine new corp stock price.  Active president must choose price for merger of majors that are both above $250. If only one major is above $250, then that becomes the price. If neither major is above $250, rightmost value that is less than sum of the two values with a $250 cap. For minors, rightmost value that is less than half the sum of the two values (i.e. the average)
- [x] Stock Exchange Part 1: For each player that owns at least two shares of combined merging corps - or controls a corporation that owns two shares of combined merging corps - beginning clockwise from player that triggered merger: 1) trade each pair of shares for one share of new corp, giving president cert if they are the first to own 20% of new corp, 2) if president share of new corp is not yet in player hands, offer them the opportunity to pay (0.5x or 1x new share price) to upgrade to president share (they will have had 2 or 3 of pre-merge combined shares). Pairs of shares in bank pool or in frozen corps are automatically traded for shares of new corp (in that order).
- [x] Stock Exchange Part 2: for each player that owns a single share of merging corps - or controls a corporation that owns a single share of merging corps - beginning clockwise from player that triggered merger: they choose between being paid 1/2 of new share price for selling outgoing share, or paying 1/2 of new share price to buy full share of new corp, or if president share of new corp is not yet in player hands, offer them the opportunity to pay to upgrade to president share by paying 1.5x new share price. Single shares of merging corps in bank pool are discarded. Single shares of merging corps in frozen corps are automatically sold for 1/2 of new share price.
- [x] If president share not in player hands, must back out of merger (throw up "dead end" and ask player to manually undo).
- [x] Any unallocated shares for new corp are placed in new corp IPO.
- [x] Swap out old tokens for new. New corp can have 5 tokens max. Ask active president to select which cities that have one or two of the outgoing tokens to place new tokens in. Cannot have multiple tokens in same city. Must have at least one token in a city.
- [x] Outgoing historical corps are removed from game. Outgoing non-historical corps are available for starting.
- [x] New corp will operate starting with following OR
- [x] Cannot merge if it will create a circular chain of ownership

**Transforming Minors** 
- [x] Only non-historical or post phase-4 minors (not exactly the same as for a merger)
- [x] New step: transform. Uses "merge" action since need to pick corp to transform into.
- [x] Automatic: move assets over. Substitute tokens.
- [x] Automatic: exchange player, corporate and share pool certs (20% for 40% share, 10% for 20% shares)
- [x] New step: offer_shares - to existing shareholders (except share pool and frozen corps, same order as merger exchange). Call @game method to trigger offer_shares - needs to see if entity can buy a share before enqueing.
- [x] repeat offer_shares - just to players (same order as above)
- [x] Remaining shares (if any) should be in IPO
- [x] Expanded buy_new_tokens step - 2nd token (if not present) is mandatory at $50 - possibly requiring EMR. Additional tokens (to total of 5) are $100 (no EMR allowed)
- [x] Automatic - new corp takes same place as outgoing corp in stock market
- [x] Automatic - new corp replaces outgoing corp in @round.entities[]
- [x] Automatic - restart non-historical outgoing corp, close historical outgoing

**Tuscan Merge**
- [x] Start of phase 4
- [x] If only one or none of SFLP, SFMA and SSFL are active, nothing happens. The SFLi is removed from the game. 
- [x] If SFLP and SFMA are active, but not SSFL, they are merged to form the SFLi
- [x] If only one of SFLP and SFMA, as well as SSFL, first the minor is transformed into a major (pick one, only transfer tokens, don't buy more yet), then the major is merged with the SSFL to form the SFLi.
- [x] If all three are active, first SFLP and SFMA are merged into a major (pick one, only transfer tokens, don't buy more yet), then the major is merged with the SSFL to form the SFLi.
- [x] All merges are performed by the president of the corp with the highest market price
- [x] When the merger to form the SFLi is done, it will have five tokens. At least one SFLi token must be in a city on the map. A token in Pisa or Firense must be left on the map. Otherwise, normal token removal occurs.
- [x] SFLi will operate in the current round, unless SFLP, SFMA or SSFL operated this round.
- [x] SFLi will lose the rest of its OR turn if SFLP, SFMA or SSFL triggered the phase change.
- [x] Deal with case where SFLP, SFMA, or SSFL is frozen (unlikely, but possible) by having the player that triggered the phase change act as their respective presidents.

**Ferdinandea Secession**
- [x] Start of phase 4, after regions change
- [x] If connection between Milano and Venezia, nothing happens
- [x] Bring in the SB and SFL at same share price as IRSFF (SB on top of SFL)
- [x] Placed IRSFF tokens in Austria go to the SB, other placed tokens to SFL
- [x] Unplaced IRSFF tokens split between SB and SFL, with SB getting the odd token
- [x] IRSFF trains from best to worst alternately given to SB and SFL, with SB getting the best
- [x] 10% of IRSFF cash goes to SFL (rounded down), remainder to SB.
- [x] IRSFF treasury shares from highest to lowest certificate price alternately given to SB and SFL, with SB getting the highest
- [x] Each shareholder, starting with the president of IRSFF and in same order as for mergers (share_pool and frozen corps last), exchanges sets of 20% (either a pair of shares or the president share) of IRSFF shares for one SB and one SFL share - except the first pair is exchanged for either the SB or SFL president cert (player choice) and the second pair (if there is one) for the other. If no other player has another pair of IRSFF shares, the other will be placed in the share_pool.
- [x] Each shareholder that has a single share, in the same order as before, exchanges that share for either a SB or SFL share (subject to availability, player choice). IRSFF president chooses if share_pool gets SB or SFL. Frozen corps get SB if available or SFL otherwise.
- [x] SB and SFL will now have presidents or be frozen
- [x] Every shareholder will be given the opportunity to buy a share (like in transforming) of SB.
- [x] The president of the SB can elect (choose) to start another round of offers and my continue to do so as long as there are IPO shares available.
- [x] Every shareholder will be given the opportunity to buy a share of SFL. If SFL is frozen, starts with IRSFF president
- [x] The SFL (or IRSFF) president can elect (choose) to start another round of offers and may continue to do so as long as there are IPO shares available.
- [x] If either the SFL or SB has only one token, a 2nd is given to it for free.
- [x] The president of the SB can choose to buy 1-3 tokens with SB cash for $50 each. Ditto for SFL.
- [x] IRSFF is closed
- [x] If IRSFF has not operated, SB and SFL will operate this round (add to round entitites). If IRSFF triggered the phase change, it will forfeit the rest of the OR.
- [x] Deal with case where IRSFF is frozen (unlikely, but possible) by having the president that triggered the phase change, make decisions as the president of the IRSFF.

**Emergency Money Raising**
- [x] EMR for buying tokens when starting a corporation is automatic - it can issue shares to cover in all cases
- [x] EMR for buying additional tokens when converting a minor to a major will require an explicit EMR step. Unclear what to do if this results in player bankruptcy.
- [x] Start with currently operating corp, then move through chain of control until a player is reached
- [x] First, use cash - automatic
- [x] Then, sell non-IPO shares - choice if more than enough to cover, automatic otherwise
- [x] Then, sell IPO shares (corps only) - automatic
- [x] Sales of stock are at half price - rounding down the total (not each individual certificate)
- [x] Sales of stock during EMR count as sale for determining legality of buying shares during a corporate financial operation
- [x] If a bankrupting player owns the concession (before Phase 4) for a historical corporation, the concession is moved to the bank and the president's share is moved to the share pool (at half price, as usual for EMR) and the corp is frozen. Until phase 4, the concession and president's share must be purchased together. Upon phase 4, the concession removed as usual and any player with 2 or more shares will become president, with ties broken starting with the player that triggered the phase change.
- [x] This needs to be incorporated into the buy_train and the buy_new_tokens step.

**Handling Frozen Corporations**
- [x] Skip OR steps, except for moving stock 2x to left and selling any shares of controlling corps for circular ownership corps
- [x] Probably? For circular ownership corps, stock in all controlling corps is sold, not just enough to break the chain
- [x] Probably? This continues for all the circular corps as they run not just the first, even if the first breaks the chain. Need to "capture" list of circular corps at start of OR.
- [x] At end of Stock round, drop price 2x down if any shares in market
- [x] Display frozen corps on Entities tab

**1841Lite**
- [x] Modified map. TBD.
- [x] Initial bank is $10900. Total distributed between players is $2520
- [x] No concessions 2, 4, and 6
- [x] No SFLP, SFLi, SFMA and SSFL corporations
- [x] No Tuscan Merge
- [x] 2x2T, 1x3T, 1x4T, and 1x5T are removed from the train roster

**Buying Stock/Concessions Directly from Players**
- [x] buying a concession from bank or player will come with pres cert if concession is active

**End Game**
- [x] Add new condition: if every red region connects to a city
- [x] Will need modify and use DestinationConnection action (via auto_actions in Destinate OR step) triggered after track lays
- [x] Add new end game timing: end of corporation turn in OR
- [x] Max stock price triggers immediately in SR and at end of corp turn in OR
- [x] All red regions connected trigger at end of corp turn
- [x] Running out of money ends game at end of OR
- [x] One player left due to bankruptcy triggers immediately

**Optional Version 1**
- [x] Modify corps
- [x] Modify companies
- [x] Modify tiles
- [x] Modify hexes
- [x] Remove restriction on tokening Firenze
- [x] New auction: Blind Auction

**Spreadsheet Improvements**
- [ ] Show subsidiary corporations somehow
- [ ] Show corporations controlled by player somehow

**Comment by KlausKiermeier** [comment]
Hey Steve, I do hope there will be a variant with the first-edition rules and map. The Deep Thought Games edition does contain this alternative version.

**Comment by roseundy** [comment]
First I will have to stop typing in 1861 when I mean 1841.

**Comment by KlausKiermeier** [comment]
True!

**Comment by roseundy** [comment]
I'm to be honest and say that my main focus will be the basic 2nd edition rules, which will be an enormous effort on its own. I (or others) can add optional rules once that is done.

**Comment by l33twash0r** [comment]
What about tile laying rules? Are they gonna require some other changes? Not sure if any of the current games check track lays from token. Affects mostly one company that has tokens in two different locations.
Major companies can lay more than one track:
• Phase 2: up to one per token. Each tile must be laid from a different token. Must be two different hexes i.e. cannot lay & upgrade
• Phase 3 & 4: Same as in phase 2 but max 2
• Phases 5-8: 1 tile only

**Comment by roseundy** [comment]
Thanks. I haven't finished with this list. Also, this was never meant to be a complete list of work to do and I haven't really started on the checklist.

**Comment by roseundy** [comment]
Yes, tracking tile lays from a token is mostly new. 21Moon has something similar, but there are only two possible graphs to check in that game.

**Comment by roseundy** [comment]
@KlausKiermeier you will be happy to know that my code supports first-edition (version 1) rules. Or at least it tries to.

**Comment by KlausKiermeier** [comment]
Hey Steve, yes I am happy :-)
Thank you very much!
And I already started a version-1 game...

**Comment by roseundy** [comment]
Sadly, looking at your game, I found a bug: my player order calculation after the blind auction needs to be reversed. I have a typo in my code that causes this.
